test-data/unit/check-classes.test

```diff
@@ -5117,3 +5117,295 @@ class C:
+import a
+import c
+class Sub(a.Base):
+    @c.deco
+    def x(self) -> int: pass  # E: Signature of "x" incompatible with supertype "Base"
+
+[file a.py]
+import b
+import c
+class Base:
+    def __init__(self) -> None:
+        self.x = 1
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicOverrideCheckedDecoratorDeferred]
+import a
+[file b.py]
+import a
+import c
+class Sub(a.Base):
+    @c.deco
+    def x(self) -> int: pass  # E: Signature of "x" incompatible with supertype "Base"
+
+[file a.py]
+import b
+import c
+class Base:
+    def __init__(self) -> None:
+        self.x = f()
+
+def f() -> int: ...
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicOverrideAnyDecoratorDeferred]
+import a
+[file b.py]
+import a
+import c
+class Sub(a.Base):
+    @c.deco
+    def x(self) -> int: pass
+
+[file a.py]
+import b
+import c
+class Base:
+    def __init__(self) -> None:
+        self.x = f()
+
+def f() -> int: ...
+[file c.py]
+from typing import Any, Callable
+def deco(f: Callable[..., Any]) -> Any: ...
+[out]
+
+[case testCyclicDecoratorDoubleDeferred]
+import b
+[file a.py]
+import b
+import c
+
+class A(b.B):
+    @c.deco
+    def meth(self) -> int:
+        reveal_type(other.x)  # E: Revealed type is 'builtins.int'
+        return 0
+
+other: Other
+class Other:
+    def __init__(self) -> None:
+        self.x = f()
+def f() -> int: ...
+[file b.py]
+import a
+import c
+
+class B:
+    @c.deco
+    def meth(self) -> int:
+        pass
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicDecoratorSuper]
+import b
+[file a.py]
+import b
+import c
+
+class A(b.B):
+    @c.deco
+    def meth(self) -> int:
+        y = super().meth()
+        reveal_type(y)  # E: Revealed type is 'Tuple[builtins.int*, builtins.int]'
+        return 0
+[file b.py]
+import a
+import c
+
+class B:
+    @c.deco
+    def meth(self) -> int:
+        pass
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicDecoratorBothDeferred]
+import b
+[file a.py]
+import b
+import c
+
+class A(b.B):
+    @c.deco
+    def meth(self) -> int:
+        pass
+[file b.py]
+import a
+import c
+
+class B:
+    @c.deco
+    def meth(self) -> int:
+        reveal_type(other.x)  # E: Revealed type is 'builtins.int'
+        return 0
+
+other: Other
+class Other:
+    def __init__(self) -> None:
+        self.x = f()
+def f() -> int: ...
+[file c.py]
+from typing import TypeVar, Tuple, Callable
+T = TypeVar('T')
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicDecoratorSuperDeferred]
```

Comment:
Did you verify that these test cases trigger deferral? Using `from ... import` can be used to make it more explicit in which order modules are processed by mypy.

Reply:
I actually checked only first two tests I added, I will also double-check rest just in case.

Reply:
They all get deferred expected number of times (1, 2, or 3), I switched to `from ... import` in half of the tests anyway.